pkg/expapi/types.go

```diff
@@ -61,3 +64,49 @@ type Scale struct {
 type ReplicationControllerDummy struct {
 	api.TypeMeta `json:",inline"`
 }
+
+// SubresourceReference contains enough information to let you inspect or modify the referred subresource.
+type SubresourceReference struct {
+	Kind        string `json:"kind,omitempty"`
+	Namespace   string `json:"namespace,omitempty"`
+	Name        string `json:"name,omitempty"`
+	APIVersion  string `json:"apiVersion,omitempty"`
+	Subresource string `json:"subresource,omitempty"`
+}
+
+// ResourceTarget is an object for specifying target average resource consumption of a particular resource.
+type ResourceTarget struct {
+	Resource api.ResourceName  `json:"resource,omitempty"`
+	Quantity resource.Quantity `json:"quantity,omitempty"`
+}
+
+// HorizontalPodAutoscalerSpec is the specification of a horizontal pod autoscaler.
+type HorizontalPodAutoscalerSpec struct {
+	// Target is a reference to Scale subresource. HorizontalPodAutoscaler will learn the current resource consumption from its status,
+	// and will set the desired number of pods by modyfying its spec.
+	Target *SubresourceReference `json:"target"`
+	// MinCount is the lower limit for the desired number of pods that can be set by the autoscaler.
+	MinCount int `json:"minCount"`
+	// MaxCount is the upper limit for the desired number of pods that can be set by the autoscaler. It cannot be smaller than MinCount.
+	MaxCount int `json:"maxCount"`
+	// TargetConsumption is the target average consumption of the given resource that the autoscaler will try to maintain by adjusting the desired number of pods.
+	// Currently two types of resources are supported: "cpu" and "memory".
+	TargetConsumption ResourceTarget `json:"targetConsumption"`
```

Comment:
"consumption" is very vague. I guess that has some benefit, in that it allows flexibility in how you define consumption. However, once you do the implementation, please add to the comment here explaining exactly where the value comes from.